driver/chain_syncer/calldata/syncer.go

```diff
@@ -223,6 +230,136 @@ func (s *Syncer) onBlockProposed(
 	return nil
 }
 
+// handleReorg detects reorg and rewinds the chain by 1 until we find a block that is still in the chain,
+// then inserts that chain as the new head.
+func (s *Syncer) handleReorg(ctx context.Context, event *bindings.TaikoL1ClientBlockProposed) error {
+	log.Info(
+		"Reorg detected",
+		"L1Height", event.Raw.BlockNumber,
+		"L1Hash", event.Raw.BlockHash,
+		"BlockID", event.Id,
+		"Removed", event.Raw.Removed,
+	)
+
+	// rewind chain by 1 until we find a block that is still in the chain
+	var lastKnownGoodBlockId *big.Int
+	var blockId *big.Int = s.lastInsertedBlockID
+	var block *types.Block
+	var err error
+
+	for {
+		if blockId.Cmp(big.NewInt(0)) == 0 {
+			lastKnownGoodBlockId = new(big.Int).SetUint64(0)
+			break
+		}
+
+		block, err = s.rpc.L2.BlockByNumber(ctx, blockId)
+		if err != nil && !errors.Is(err, ethereum.NotFound) {
+			return err
+		}
+
+		if block != nil {
+			// block exists, we can rewind to this block
```

Comment:
do we need to check that this blockID is also smaller than `TaikoL1.state.numBlocks`?